### ISSUE TITLE
FFSuperHirn segfaults on build machines

### ISSUE THREAD

**Issue by jpfeuffer** [OpenMS/OpenMS#2549]
I wanted to fix the ...XMLFile_tests from e.g. http://cdash.openms.de/testDetails.php?test=39575898&build=220777
 where I wanted to do a "make install" to overwrite the global outdated OpenMS/share.

However, FFSuperhirn segfaults when you execute the binary (e.g. when it should print the parameter documentation etc).

MacOSX Lion. gcc 5.2.

gdb StackTrace:
```Program received signal EXC_BAD_ACCESS, Could not access memory.
Reason: KERN_INVALID_ADDRESS at address: 0x0000000000000000
0x00000001001376a3 in xercesc_3_1::XMLString::replicate ()
(gdb) backtrace
#0  0x00000001001376a3 in xercesc_3_1::XMLString::replicate ()
#1  0x000000010003237b in xercesc_3_1::ENameMap::ENameMap ()
#2  0x00000001000329e9 in xercesc_3_1::ENameMapFor<xercesc_3_1::XMLChTranscoder>::ENameMapFor ()
#3  0x000000010192c1ff in xercesc_3_1::XMLTransService::initTransService ()
#4  0x000000010190d9a3 in xercesc_3_1::XMLPlatformUtils::Initialize ()
#5  0x00000001012a84b4 in OpenMS::Internal::XMLFile::parse_ (this=0x7fff5fbfc940, filename=@0x7fff5fbfca20, handler=0x7fff5fbfc5e0) at /Users/aiche/NIGHTLY/HEAD/src/openms/source/FORMAT/XMLFile.cpp:118
#6  0x00000001012676df in OpenMS::ParamXMLFile::load (this=0x7fff5fbfc940, filename=@0x7fff5fbfca20, param=@0x7fff5fbfc9b0) at /Users/aiche/NIGHTLY/HEAD/src/openms/source/FORMAT/ParamXMLFile.cpp:366
#7  0x0000000100fbaa2a in OpenMS::ResidueDB::readResiduesFromFile_ (this=0x10603cc00, file_name=@0x7fff5fbfcd00) at /Users/aiche/NIGHTLY/HEAD/src/openms/source/CHEMISTRY/ResidueDB.cpp:196
#8  0x0000000100fb9ae6 in OpenMS::ResidueDB::ResidueDB (this=0x10603cc00) at /Users/aiche/NIGHTLY/HEAD/src/openms/source/CHEMISTRY/ResidueDB.cpp:57
#9  0x0000000100fa0e49 in OpenMS::ResidueDB::getInstance () at ResidueDB.h:78
#10 0x0000000100fd3508 in OpenMS::SvmTheoreticalSpectrumGenerator::initializeMaps_ () at /Users/aiche/NIGHTLY/HEAD/src/openms/source/CHEMISTRY/SvmTheoreticalSpectrumGenerator.cpp:97
#11 0x0000000100fd71f1 in OpenMS::SvmTheoreticalSpectrumGenerator::SvmTheoreticalSpectrumGenerator (this=0x1025fe5c0) at /Users/aiche/NIGHTLY/HEAD/src/openms/source/CHEMISTRY/SvmTheoreticalSpectrumGenerator.cpp:185
#12 0x0000000100fe10f5 in __static_initialization_and_destruction_0 (__initialize_p=1, __priority=65535) at /Users/aiche/NIGHTLY/HEAD/src/openms/source/CHEMISTRY/SvmTheoreticalSpectrumGenerator.cpp:66
#13 0x0000000100fe12df in _GLOBAL__sub_I_SvmTheoreticalSpectrumGenerator.cpp () at /Users/aiche/NIGHTLY/HEAD/src/openms/source/CHEMISTRY/SvmTheoreticalSpectrumGenerator.cpp:1055
#14 0x00007fff5fc13378 in __dyld__ZN16ImageLoaderMachO18doModInitFunctionsERKN11ImageLoader11LinkContextE ()
#15 0x00007fff5fc13762 in __dyld__ZN16ImageLoaderMachO16doInitializationERKN11ImageLoader11LinkContextE ()
#16 0x00007fff5fc1006e in __dyld__ZN11ImageLoader23recursiveInitializationERKNS_11LinkContextEjRNS_21InitializerTimingListE ()
#17 0x00007fff5fc0ffc4 in __dyld__ZN11ImageLoader23recursiveInitializationERKNS_11LinkContextEjRNS_21InitializerTimingListE ()
#18 0x00007fff5fc0ffc4 in __dyld__ZN11ImageLoader23recursiveInitializationERKNS_11LinkContextEjRNS_21InitializerTimingListE ()
#19 0x00007fff5fc0feba in __dyld__ZN11ImageLoader15runInitializersERKNS_11LinkContextERNS_21InitializerTimingListE ()
#20 0x00007fff5fc01fc0 in __dyld__ZN4dyld24initializeMainExecutableEv ()
#21 0x00007fff5fc05b04 in __dyld__ZN4dyld5_mainEPK12macho_headermiPPKcS5_S5_Pm ()
#22 0x00007fff5fc01397 in __dyld__ZN13dyldbootstrap5startEPK12macho_headeriPPKclS2_Pm ()
#23 0x00007fff5fc0105e in __dyld__dyld_start ()
```

Any ideas?

**Comment by jpfeuffer** [comment]
A thing that I found when googling for the initialize() method of xerces:
When do we actually call terminate() for a generic XMLFile? We do in the MZIdentMLDOMHandler but I did not find it anywhere else. Probably will not solve the problem, though.

**Comment by jpfeuffer** [comment]
I give up. Could basically be anything on that machine with its dozens of compilers and hundreds of different installs of OpenMS.

**Comment by cbielow** [comment]
this looks like a wrong dynamic lib being used. Which libs are dynamic on MacOS?

**Comment by hendrikweisser** [comment]
See also here: https://github.com/OpenMS/OpenMS/issues/2094#issuecomment-239151573
(I've had this problem for a long time on my Mac.)

**Comment by jpfeuffer** [comment]
Thanks for your input guys. I agree that it might be a linking issue.
Xerces was built static and the correct version from the gcc-contrib is included according to the cmakefiles. Dynamic are OpenMS/Swath and some more. I will maybe have a short look again.

**Comment by jpfeuffer** [comment]
The bad thing is the external library status of Superhirn. Since it has its own Cmake files it tries to find its own libraries again and is prone to choose wrong ones.
Unfortunately 1 hour looking at commands in Cmakefiles did not reveal an obvious error.

**Comment by cbielow** [comment]
have you tried looking at the executable's linked libs directly? `otool -L <path-to-binary>` and compare FFSH vs. something that runs?

**Comment by jpfeuffer** [comment]
Got it! otool was actually the same. But I concentrated on xerces since this is where it fails. The problem was that we include xerces statically in the dynamic libOpenMS, however somehow CMake links xerces into libSuperHirn again and this (I think) led to the crash. Not sure why though.
I went to the link.txt in the CMakeFiles of SuperHirn and removed xerces-c.a, rebuilt and executed the binary again (dynamically linking to the new library). And it works.
I think a full solution would be to not link static libraries of OpenMS into thirdparty libraries again?

**Comment by jpfeuffer** [comment]
Yes. I can reproduce the behaviour.

**Comment by jpfeuffer** [comment]
http://stackoverflow.com/questions/31503597/cmake-link-a-shared-library-to-static-libraries
https://cmake.org/pipermail/cmake/2016-May/063400.html
This could be helpful in implementing this.

**Comment by hroest** [comment]
> The bad thing is the external library status of Superhirn. Since it has its own Cmake files it tries to find its own libraries again and is prone to choose wrong ones.

@jpfeuffer I believe this was done as a "test-case" for modularizing OpenMS with the idea that with the infrastructure in place we could later on create multiple smaller independent libraries using the same approach. This has not happened so far.

However, I agree that maybe having Superhin choose its own libraries is not the best approach, but if the search routines are the same then also the same library should be chosen, right?

PS: This may be part of the explanation: http://stackoverflow.com/questions/31209693/static-library-linked-two-times

> This issue shows up when linking statically the standard library in a program and a library using that, when you get two default allocators and if you call new on a library and delete on the other, the object will leak on the new side and the heap is likely to become corrupted on the delete side.

so basically calling a delete on the wrong memory address leading to a segfault?

**Comment by jpfeuffer** [comment]
aha. Yes makes sense in the context of xerces and its intialization and memory handling routines (according to my very basic knowledge acquired yesterday).

**Comment by hroest** [comment]
so I wonder why this only happens on OSX, is this the only system we statically link xerces?

**Comment by jpfeuffer** [comment]
Sorry for editing. I was wondering, too.

**Comment by hroest** [comment]
@jpfeuffer is this resolved now? can we close this?

**Comment by jpfeuffer** [comment]
No not yet. I would like to have a general solution in the build system. Did not get to it yet.

**Comment by hroest** [comment]
I think there are two things here happening:

- the idea that we can modularize the library into multiple .so files
- the idea that we could have independent CMake files so that you could build each library independently and each one would discover the dependencies by itself

I think the first idea makes sense, the second not so much (e.g. when would you ever just build libSuperhirn.so by itself?). So maybe we could do some of the dependency discovery in a shared CMakeList.txt and then propagate the variables down to each individual build?

**Comment by hroest** [comment]
@jpfeuffer I would still opt to close this issue as the issue itself is solved and open a new one with the ideas how to improve the build system. agree?

**Comment by jpfeuffer** [comment]
The important thing is to make sure that static libraries are not linked twice if you already link against a dynamic library that included them. But yes, the bug here was identified and a solution was found that is a bit more general than this single issue. I am fine with closing this and the related ones in favor of a new general "Build System type" one.

**Comment by jpfeuffer** [comment]
Seems to be solved, too.
